src/backend/distributed/executor/executor.c

```diff
@@ -681,6 +723,12 @@ AssignTasksToConnections(DistributedExecution *execution)
 				{
 					workerPool->idleConnectionCount++;
 				}
+
+				/* we don't want any connection to be used by any other session */
+				if (ForceMaxQueryParallelization)
+				{
+					ClaimConnectionExclusively(connection);
```

Comment:
This seems a little off. The preferred way of getting a new connection for every placement is to use the `CONNECTION_PER_PLACEMENT` flag in `StartPlacementListConnection`. That also ensures that, if we modify shard 1 and then query shard 1 and 2,  we avoid using the connection that modified shard 1 to read from shard 2, which would leave shard 1 without a suitable connection.